pkg/kv/store.go

```diff
@@ -0,0 +1,111 @@
+package kv
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var (
+	ErrUnknownDriver       = errors.New("unknown driver")
+	ErrDriverConfiguration = errors.New("driver configuration")
+	ErrOperationFailed     = errors.New("operation failed")
+	ErrConnectFailed       = errors.New("connect failed")
+	ErrSetupFailed         = errors.New("setup failed")
+	ErrMissingKey          = errors.New("missing key")
+	ErrMissingValue        = errors.New("missing value")
+	ErrNotFound            = errors.New("not found")
+	ErrClosedEntries       = errors.New("closed entries")
+	ErrPredicateFailed     = errors.New("predicate failed")
+)
+
+// Driver define the root interface to access the kv database.
+// Each kv provider implements a driver which provides Store instance.
+type Driver interface {
+	// Open access to the database store. migrate kv if needed.
+	Open(ctx context.Context, dsn string) (Store, error)
+}
+
+type Store interface {
+	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
+	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Set stores the given value, overwriting an existing value if one exists
+	Set(ctx context.Context, key, value []byte) error
+
+	// SetIf returns an ErrPredicateFailed error if the key with valuePredicate passed
+	//  doesn't match the currently stored value. SetIf is a simple compare-and-swap operator:
+	//  valuePredicate is either the existing value, or nil for no key.
+	//  this is intentionally simplistic: we can model a better abstraction on top, keeping this interface simple for implementors
+	SetIf(ctx context.Context, key, value, valuePredicate []byte) error
```

Comment:
So if valuePredicate is nil this is the equivalent of 'set if not exist'?

Reply:
yes